syncstorage/storage/memcached.py

```diff
@@ -369,17 +369,22 @@ def valid_batch(self, userid, collection, batchid):
     def append_items_to_batch(self, userid, collection, batchid, items):
         """Appends items to the pending batch."""
         colmgr = self._get_collection_manager(collection)
-        ts = colmgr.append_items_to_batch(userid, batchid, items)
-        return ts
+        with self._mark_collection_dirty(userid, collection) as update:
+            ts = colmgr.append_items_to_batch(userid, batchid, items)
+            # Account for the size of the new items as they come in,
+            # since that's the only opportunity we have to see them.
+            # Don't update the timestamp yet though, as they're not committed.
+            size = sum(len(item.get("payload", "")) for item in items)
+            update(size_incr=size)
+            return ts
 
     def apply_batch(self, userid, collection, batchid):
         """Applies the batch"""
         colmgr = self._get_collection_manager(collection)
-        # prevts = colmgr.get_timestamp(userid)
-        with self._mark_collection_dirty(userid, collection):
+        with self._mark_collection_dirty(userid, collection) as update:
             ts = colmgr.apply_batch(userid, batchid)
-            # Lazy update should occur on following get_item/s
-        return ts
+            update(ts, ts)
+            return ts
```

Comment:
Here, we were chosing not to update the timestamp in the `<userid>:metadata` collection, and instead to lazily fall back to reading it from the cached collection data after applying a batch.  This is the reason why we started observing this bug when the batch upload API was enabled.